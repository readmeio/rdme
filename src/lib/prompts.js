const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
