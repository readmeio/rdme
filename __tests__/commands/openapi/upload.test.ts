import nock from 'nock';
import prompts from 'prompts';
import slugify from 'slugify';
import { describe, beforeAll, beforeEach, afterEach, it, expect } from 'vitest';

import Command from '../../../src/commands/openapi/upload.js';
import petstore from '../../__fixtures__/petstore-simple-weird-version.json' with { type: 'json' };
import { getAPIv2Mock, getAPIv2MockForGHA } from '../../helpers/get-api-mock.js';
import { runCommand, type OclifOutput } from '../../helpers/oclif.js';
import { after, before } from '../../helpers/setup-gha-env.js';

const key = 'rdme_123';
const version = '1.0.0';
const filename = '__tests__/__fixtures__/petstore-simple-weird-version.json';
const fileUrl = 'https://example.com/openapi.json';
const slugifiedFilename = slugify.default(filename);

describe('rdme openapi upload', () => {
  let run: (args?: string[]) => OclifOutput;

  beforeAll(() => {
    run = runCommand(Command);
  });

  describe('flag error handling', () => {
    it('should throw if an error if both `--version` and `--useSpecVersion` flags are passed', async () => {
      const result = await run(['--useSpecVersion', '--version', version, filename, '--key', key]);
      expect(result).toMatchSnapshot();
    });
  });

  describe('given that the API definition is a local file', () => {
    it('should create a new API definition in ReadMe', async () => {
      const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
        .get(`/versions/${version}/apis`)
        .reply(200, { data: [] })
        .post(`/versions/${version}/apis`, body =>
          body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
        )
        .reply(200, {
          data: {
            upload: { status: 'done' },
            uri: `/versions/${version}/apis/${slugifiedFilename}`,
          },
        });

      const result = await run(['--version', version, filename, '--key', key]);
      expect(result).toMatchSnapshot();

      mock.done();
    });

    it('should update an existing API definition in ReadMe', async () => {
      prompts.inject([true]);

      const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
        .get(`/versions/${version}/apis`)
        .reply(200, { data: [{ filename: slugifiedFilename }] })
        .put(`/versions/1.0.0/apis/${slugifiedFilename}`, body =>
          body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
        )
        .reply(200, {
          data: {
            upload: { status: 'done' },
            uri: `/versions/${version}/apis/${slugifiedFilename}`,
          },
        });

      const result = await run(['--version', version, filename, '--key', key]);
      expect(result).toMatchSnapshot();

      mock.done();
    });

    it('should handle upload failures', async () => {
      const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
        .get(`/versions/${version}/apis`)
        .reply(200, { data: [] })
        .post(`/versions/${version}/apis`, body =>
          body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
        )
        .reply(200, {
          data: {
            upload: { status: 'fail' },
            uri: `/versions/${version}/apis/${slugifiedFilename}`,
          },
        });

      const result = await run(['--version', version, filename, '--key', key]);
      expect(result).toMatchSnapshot();

      mock.done();
    });

    describe('and the `--slug` flag is passed', () => {
      it('should use the provided slug (no file extension) as the filename', async () => {
        const customSlug = 'custom-slug';
        const customSlugWithExtension = `${customSlug}.json`;
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${version}/apis`, body =>
            body.match(`form-data; name="schema"; filename="${customSlugWithExtension}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'done' },
              uri: `/versions/${version}/apis/${customSlugWithExtension}`,
            },
          });

        const result = await run(['--version', version, filename, '--key', key, '--slug', customSlug]);
        expect(result).toMatchSnapshot();

        mock.done();
      });

      it('should use the provided slug (includes file extension) as the filename', async () => {
        const customSlug = 'custom-slug.json';
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${version}/apis`, body => body.match(`form-data; name="schema"; filename="${customSlug}"`))
          .reply(200, {
            data: {
              upload: { status: 'done' },
              uri: `/versions/${version}/apis/${customSlug}`,
            },
          });

        const result = await run(['--version', version, filename, '--key', key, '--slug', customSlug]);
        expect(result).toMatchSnapshot();

        mock.done();
      });

      it('should handle a slug with an invalid file extension', async () => {
        const customSlug = 'custom-slug.yikes';

        const result = await run(['--version', version, filename, '--key', key, '--slug', customSlug]);
        expect(result).toMatchSnapshot();
      });

      it('should handle a slug with a valid but mismatching file extension', async () => {
        const customSlug = 'custom-slug.yml';

        const result = await run(['--version', version, filename, '--key', key, '--slug', customSlug]);
        expect(result).toMatchSnapshot();
      });
    });

    describe('and the upload status initially is a pending state', () => {
      it('should poll the API until the upload is complete', async () => {
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${version}/apis`, body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'pending' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          })
          .get(`/versions/${version}/apis/${slugifiedFilename}`)
          .times(9)
          .reply(200, {
            data: {
              upload: { status: 'pending' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          })
          .get(`/versions/${version}/apis/${slugifiedFilename}`)
          .reply(200, {
            data: {
              upload: { status: 'done' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          });

        const result = await run(['--version', version, filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });

      it('should poll the API and handle timeouts', async () => {
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${version}/apis`, body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'pending' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          })
          .get(`/versions/${version}/apis/${slugifiedFilename}`)
          .times(10)
          .reply(200, {
            data: {
              upload: { status: 'pending' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          });

        const result = await run(['--version', version, filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });

      it('should poll the API once and handle a failure state with a 4xx', async () => {
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${version}/apis`, body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'pending' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          })
          .get(`/versions/${version}/apis/${slugifiedFilename}`)
          .reply(400);

        const result = await run(['--version', version, filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });

      it('should poll the API once and handle an unexpected state with a 2xx', async () => {
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${version}/apis`, body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'pending' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          })
          .get(`/versions/${version}/apis/${slugifiedFilename}`)
          .reply(200, {
            data: {
              upload: { status: 'something-unexpected' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          });

        const result = await run(['--version', version, filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });
    });

    describe('and the command is being run in a CI environment', () => {
      beforeEach(before);

      afterEach(after);

      it('should overwrite an existing API definition without asking for confirmation', async () => {
        const mock = getAPIv2MockForGHA({ authorization: `Bearer ${key}` })
          .get(`/versions/${version}/apis`)
          .reply(200, { data: [{ filename: slugifiedFilename }] })
          .put(`/versions/1.0.0/apis/${slugifiedFilename}`, body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'done' },
              uri: `/versions/${version}/apis/${slugifiedFilename}`,
            },
          });

        const result = await run(['--version', version, filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });
    });

    describe('given that the `--version` flag is not set', () => {
      it('should default to the `stable` version', async () => {
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get('/versions/stable/apis')
          .reply(200, { data: [] })
          .post('/versions/stable/apis', body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'done' },
              uri: `/versions/stable/apis/${slugifiedFilename}`,
            },
          });

        const result = await run([filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });

      it('should use the version from the spec file if --`useSpecVersion` is passed', async () => {
        const altVersion = '1.2.3';
        const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
          .get(`/versions/${altVersion}/apis`)
          .reply(200, { data: [] })
          .post(`/versions/${altVersion}/apis`, body =>
            body.match(`form-data; name="schema"; filename="${slugifiedFilename}"`),
          )
          .reply(200, {
            data: {
              upload: { status: 'done' },
              uri: `/versions/${altVersion}/apis/${slugifiedFilename}`,
            },
          });

        const result = await run(['--useSpecVersion', filename, '--key', key]);
        expect(result).toMatchSnapshot();
        mock.done();
      });
    });
  });

  describe('given that the API definition is a URL', () => {
    it('should create a new API definition in ReadMe', async () => {
      const fileMock = nock('https://example.com').get('/openapi.json').reply(200, petstore);

      const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
        .get(`/versions/${version}/apis`)
        .reply(200, {})
        .post(`/versions/${version}/apis`, body => body.match(`form-data; name="url"\r\n\r\n${fileUrl}`))
        .reply(200, {
          data: {
            upload: { status: 'done' },
            uri: `/versions/${version}/apis/openapi.json`,
          },
        });

      const result = await run(['--version', version, fileUrl, '--key', key]);
      expect(result).toMatchSnapshot();
      fileMock.done();
      mock.done();
    });

    it('should update an existing API definition in ReadMe', async () => {
      prompts.inject([true]);

      const fileMock = nock('https://example.com').get('/openapi.json').reply(200, petstore);

      const mock = getAPIv2Mock({ authorization: `Bearer ${key}` })
        .get(`/versions/${version}/apis`)
        .reply(200, { data: [{ filename: 'openapi.json' }] })
        .put('/versions/1.0.0/apis/openapi.json', body => body.match(`form-data; name="url"\r\n\r\n${fileUrl}`))
        .reply(200, {
          data: {
            upload: { status: 'done' },
            uri: `/versions/${version}/apis/openapi.json`,
          },
        });

      const result = await run(['--version', version, fileUrl, '--key', key]);
      expect(result).toMatchSnapshot();
      fileMock.done();
      mock.done();
    });

    it('should handle issues fetching from the URL', async () => {
      const fileMock = nock('https://example.com').get('/openapi.json').reply(400, {});

      const result = await run(['--version', version, fileUrl, '--key', key]);
      expect(result).toMatchSnapshot();
      fileMock.done();
    });
  });
});
