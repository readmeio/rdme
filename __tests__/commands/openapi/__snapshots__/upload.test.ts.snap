// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rdme openapi upload > flag error handling > should throw if an error if both \`--branch\` and \`--useSpecVersion\` flags are passed 1`] = `
{
  "error": [Error: The following error occurred:
  --branch=1.0.0 cannot also be provided when using --useSpecVersion
See more help with --help],
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme openapi upload > flag error handling > should throw if an error if both \`--slug\` and \`--legacy-id\` flags are passed 1`] = `
{
  "error": [Error: The following error occurred:
  --slug=custom-slug cannot also be provided when using --legacy-id
See more help with --help],
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the "--legacy-id" flag is passed > should error if no matching API definition found 1`] = `
{
  "error": [Error: No API definition found with legacy ID 1234567890.],
  "stderr": " â€º   Warning: The "legacy-id" flag has been deprecated.
- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the "--legacy-id" flag is passed > should update an existing spec with matching legacy_id 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/legacy-spec.json",
  },
  "stderr": " â€º   Warning: The "legacy-id" flag has been deprecated.
- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
- Updating your API definition to ReadMe...
âœ” Updating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (legacy-spec.json) was successfully updated in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a URL > should create a new API definition in ReadMe 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/openapi.json",
  },
  "stderr": "- Validating the API definition located at https://example.com/openapi.json...
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (openapi.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a URL > should handle issues fetching from the URL 1`] = `
{
  "error": [Error: Unknown file detected.],
  "stderr": "- Validating the API definition located at https://example.com/openapi.json...
âœ– Validating the API definition located at https://example.com/openapi.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a URL > should update an existing API definition in ReadMe 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/openapi.json",
  },
  "stderr": "- Validating the API definition located at https://example.com/openapi.json...
- Updating your API definition to ReadMe...
âœ” Updating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (openapi.json) was successfully updated in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the \`--slug\` flag is passed > should handle a slug with a valid but mismatching file extension 1`] = `
{
  "error": [Error: Please provide a valid file extension that matches the extension on the file you provided. Must be \`.json\`, \`.yaml\`, or \`.yml\`.],
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the \`--slug\` flag is passed > should handle a slug with an invalid file extension 1`] = `
{
  "error": [Error: Please provide a valid file extension that matches the extension on the file you provided. Must be \`.json\`, \`.yaml\`, or \`.yml\`.],
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the \`--slug\` flag is passed > should use the provided slug (includes file extension) as the filename 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/custom-slug.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (custom-slug.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the \`--slug\` flag is passed > should use the provided slug (no file extension) as the filename 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/custom-slug.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (custom-slug.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the command is being run in a CI environment > should overwrite an existing API definition without asking for confirmation 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
- Updating your API definition to ReadMe...
âœ” Updating your API definition to ReadMe... done!
",
  "stdout": "::warning::The slug of your API Definition will be set to __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This slug is not visible to your end users. To set this slug to something else, use the \`--slug\` flag.
ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully updated in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the upload status initially is a pending state > should poll the API and handle timeouts 1`] = `
{
  "error": [Error: Sorry, this upload timed out. Please try again later.],
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the upload status initially is a pending state > should poll the API once and handle a failure state with a 4xx 1`] = `
{
  "error": [Error: The ReadMe API responded with an unexpected error. Please try again and if this issue persists, get in touch with us at support@readme.io.],
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the upload status initially is a pending state > should poll the API once and handle an unexpected state with a 2xx 1`] = `
{
  "error": [Error: Your API definition upload failed with an unexpected error. Please get in touch with us at support@readme.io.],
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ– Creating your API definition to ReadMe... uploaded but not yet processed by ReadMe. Polling for completion...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the upload status initially is a pending state > should poll the API until the update is complete 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Updating your API definition to ReadMe...
âœ” Updating your API definition to ReadMe... uploaded but not yet processed by ReadMe. Polling for completion... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully updated in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > and the upload status initially is a pending state > should poll the API until the upload is complete 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... uploaded but not yet processed by ReadMe. Polling for completion... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > given that the \`--branch\` flag is not set > should default to the \`stable\` version 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/stable/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > given that the \`--branch\` flag is not set > should use the version from the spec file if --\`useSpecVersion\` is passed 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.2.3/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > should create a new JSON API definition in ReadMe 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > should create a new JSON API definition in ReadMe with deprecated \`--version\` flag 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": " â€º   Warning: The "--version" flag has been deprecated. Use "--branch" instead.
- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > should create a new YAML API definition in ReadMe 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__postmanpetstore.collection.yaml",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/postman/petstore.collection.yaml...
 â€º   Warning: Support for Postman collections is currently experimental.
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__postmanpetstore.collection.yaml in ReadMe. This slug 
 â€º   is not visible to your end users. To set this slug to something else, use 
 â€º   the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ” Creating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__postmanpetstore.collection.yaml) was successfully created in ReadMe!
",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > should handle upload failures 1`] = `
{
  "error": [Error: Your API definition upload failed with an unexpected error. Please get in touch with us at support@readme.io.],
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Creating your API definition to ReadMe...
âœ– Creating your API definition to ReadMe...
",
  "stdout": "",
}
`;

exports[`rdme openapi upload > given that the API definition is a local file > should update an existing API definition in ReadMe 1`] = `
{
  "result": {
    "status": "done",
    "uri": "/branches/1.0.0/apis/__tests____fixtures__petstore-simple-weird-version.json",
  },
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
 â€º   Warning: The slug of your API Definition will be set to 
 â€º   __tests____fixtures__petstore-simple-weird-version.json in ReadMe. This 
 â€º   slug is not visible to your end users. To set this slug to something else,
 â€º    use the \`--slug\` flag.
- Updating your API definition to ReadMe...
âœ” Updating your API definition to ReadMe... done!
",
  "stdout": "ðŸš€ Your API definition (__tests____fixtures__petstore-simple-weird-version.json) was successfully updated in ReadMe!
",
}
`;
